daffodil-runtime1-layers/src/main/scala/org/apache/daffodil/layers/runtime1/GZipLayer.java

```diff
@@ -54,29 +51,30 @@ public static boolean fixIsNeeded() {
 
   public GZipLayer() {
     super("gzip", "urn:org.apache.daffodil.layers.gzip");
+
+    // we need a layer throw handler to convert any IOExceptions into
+    // processing errors so they can backtrack (i.e., are not fatal)
+    setLayerThrowHandler(new LayerThrowHandler() {
+      @Override
+      public void handleThrow(Throwable t) {
+        // Fuzz testing shows gzip does a good job converting all sorts of parse errors
+        // into IOException. If there were other kinds of exceptions thrown, we could also
+        // catch them here.
+        if (t instanceof IOException)
+          processingError(t);
+      }
+    });
   }
```

Comment:
I personally am not a huge fan of this approach. I get the reason that its annoying to have a bunch of try/catches all over the place, but that's kindof just that the way it goes when using JVMs. It's ugly, but people understand and are used to it. I think we're just too used to Scala where (for better or for worse) we kindof just ignore most exceptions and can't stand to see try/catch everywhere.

I'd prefer implementations handle catching errors where they are thrown and converting them to PE's themselves.

And if they want something like a general exception handler like this, they can pretty easily implement it themselves by wrapping their functions in a try/catch and call their own handleThrow function, e.g.

```java
@Override
Int read() {
  try {
    ...
  }
  catch (Exception e) {
    handleThrow(e)
  }
}
```

Scala implementations can be even less verbose, e.g.

```scala
def handleThrow[A](f: () => A) {
  try {
    f()
  } catch {
    case e: Exception => ...
  }
}

override read(): Int = handleThrow {
  ...
}
```

But I don't really think layers should that. I prefer exceptions to be handled closer to where they are thrown.

Reply:
Ok, so take a look at this same PR, all minus the last commit. That will show you what gzip looks like without the framework-supported throw handler thing. 

Note that one can write a layer that way, ignoring the throw handler feature. It will work. The throw handler is just a pre-built way to do all those try-catches in a uniform way, which is what I expect basically all layers to do. But it's not required since users are being allowed to implement a Java InputStream and OutputStream themselves in this API still. 

A last thought. One could just have a policy setting e.g.,

    setParserBacktrackOn(classOf[NullPointerException], classOf[IOException])

I was thinking more flexibility is needed, but I expect most layers would get by fine with just a policy like this.

Reply:
Yeah, the previous commit is definitely ugly, I can see why you wanted to get rid of all the try/catches.

I can't say I've ever seen a Java API that has a function like `setParserBacktrackOn` that defines a list of exceptions to treat differently, but maybe this Layer stuff is a special case, and it does feel pretty clean. I do like it better than the handler approach. It's less flexible than the handler, but I'm not sure users need that flexibility. And if they really need it, they can catch exceptions themselves and do whatever they want.

Reply:
Ok. I will create  set policy thing. How about....

      setProcessingErrorExceptions(...)

I think "set" needs to be the start of the name because the state of the layer is being changed in a way that changes the behavior. "ProcessingErrorExceptions"  seems better than "ExceptionsConvertedToProessingErrors"

Maybe avoiding varargs and collections is better. i.e., it should be one exception class at a time:

    setProcessingErrorException(classOf[NullPointerException])
    setProcessingErrorException(classOf[IOException])

When one writes try/catch, each catch handles one exception class at a time.

Reply:
Another option is to make it a getter that is overridden by the Layer and called by Daffodil, maybe  something like:

```java
@Override
List<Class<Throwable>> getProcessingErrorExceptions() {
  return Arrays.asList(
    NullPointerException.class,
    IOException.class
  );
}
```

Not sure if that's any better. I have no strong preference either way.

That said, all this extra complication is making wonder if maybe we do really want to just say "all exceptions are PE's". It feels like it could hide bugs, but it is definitely simple and clear.

Reply:
Check it out now. Latest push has setProcessingErrorException(e) call in use. I converted boundary mark also, and the daffodil-test layers also. 

With your approval I will push this back to the original PR and close this within-my-fork PR.

Reply:
Yep, seems like a reasonable approach :+1: